### PR TITLE
fix(s3stream/storage): calculate the confirm offset from unordered requests

### PR DESCRIPTION
**Pull request by Chillax-0v0**

### CODE REVIEW

File: s3stream/src/main/java/com/automq/stream/s3/S3Storage.java [hunk 1]
```diff
@@ -534,12 +549,115 @@ private void freeCache(LogCache.LogCacheBlock cacheBlock) {
...
+                assert wrapper != null;
+                if (wrapper.request == null) {
+                    // Reach the flag.
+                    break;
+                }
+                WalWriteRequest request = wrapper.request;
+                if (request.persisted && request.offset < minUnconfirmedOffset) {
+                    confirmedOffset = confirmedOffset == null ? request.offset : Math.max(confirmedOffset, request.offset);
+                }
+            }
```
Comment (Chillax-0v0) [inline s3stream/src/main/java/com/automq/stream/s3/S3Storage.java:636]: bug here: only requests before the confirm offset can be dropped

File: s3stream/src/main/java/com/automq/stream/s3/S3Storage.java [hunk 2]
```diff
@@ -534,12 +549,115 @@ private void freeCache(LogCache.LogCacheBlock cacheBlock) {
...
+                assert request.offset != NOOP_OFFSET;
+                if (request.persisted) {
+                    maxPersistedOffset = Math.max(maxPersistedOffset, request.offset);
+                } else {
+                    minUnconfirmedOffset = minUnconfirmedOffset == null ? request.offset : Math.min(minUnconfirmedOffset, request.offset);
+                }
+            }
+            assert reachFlag;
+
+            if (null == minUnconfirmedOffset) {
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/s3/S3Storage.java:618]: If all requests are persisted, then pop nothing?

File: s3stream/src/main/java/com/automq/stream/s3/S3Storage.java [hunk 3]
```diff
@@ -534,12 +549,115 @@ private void freeCache(LogCache.LogCacheBlock cacheBlock) {
...
+                // All offsets are confirmed.
+                return maxPersistedOffset;
+            }
+
+            Long confirmedOffset = null;
+            while (true) {
+                // Iterate and poll the queue to find the max offset less than minUnconfirmedOffset.
+                WalWriteRequestWrapper wrapper = queue.poll();
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/s3/S3Storage.java:626]: If the request.offset > minUnconfirmedOffset, then the request will also be poped

File: s3stream/src/main/java/com/automq/stream/s3/S3Storage.java [hunk 4]
```diff
@@ -534,12 +549,115 @@ private void freeCache(LogCache.LogCacheBlock cacheBlock) {
...
+            if (offset != NOOP_OFFSET) {
+                confirmOffset.set(offset);
+            }
+        }
+
+        /**
+         * Calculate the offset before and including which all records have been persisted.
+         * It returns {@link WALCallbackSequencer#NOOP_OFFSET} if the first record is not persisted yet.
+         */
+        synchronized private long calculate() {
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/s3/S3Storage.java:589]: 1. First, find the minUnconfirmOffset.
2. Then, iterate through all values less than minUnconfirmOffset and remove them using the iterator, while keeping track of maxSuccessiveConfirmOffset.

**Review by superhx — APPROVED** [review]

**Pull request merged**